core/src/main/java/org/apache/calcite/rel/core/RelFactories.java

```diff
@@ -549,6 +555,34 @@ public RelNode createMatch(RelNode input, RexNode pattern,
           partitionKeys, orderKeys, interval);
     }
   }
+
+  /**
+   * Can create a {@link TableFunctionScan}
+   * of the appropriate type for a rule's calling convention.
+   */
+  public interface TableFunctionScanFactory {
+    /** Creates a {@link TableFunctionScan}. */
+    RelNode createTableFunctionScan(RelOptCluster cluster,
+        List<RelNode> inputs, RexNode rexCall, Type elementType,
+        RelDataType rowType, Set<RelColumnMapping> columnMappings);
```

Comment:
Is `rowType` necessary? Is it possible to ask the table function to deduce its own row type?

Reply:
Yep, I think its rowType is same with function.